src/Sentry/MetricAggregator.cs

```diff
@@ -0,0 +1,273 @@
+    }
+
+    /// <inheritdoc cref="IMetricAggregator.Increment"/>
+    public void Increment(
+        string key,
+        double value = 1.0,
+        MeasurementUnit? unit = null,
+        IDictionary<string, string>? tags = null,
+        DateTime? timestamp = null
+        // , int stacklevel = 0 // Used for code locations
+        ) => Emit(MetricType.Counter, key, value, unit, tags, timestamp);
+
+    /// <inheritdoc cref="IMetricAggregator.Gauge"/>
+    public void Gauge(
+        string key,
+        double value = 1.0,
+        MeasurementUnit? unit = null,
+        IDictionary<string, string>? tags = null,
+        DateTime? timestamp = null
+        // , int stacklevel = 0 // Used for code locations
+    ) => Emit(MetricType.Gauge, key, value, unit, tags, timestamp);
+
+    /// <inheritdoc cref="IMetricAggregator.Distribution"/>
+    public void Distribution(
+        string key,
+        double value = 1.0,
+        MeasurementUnit? unit = null,
+        IDictionary<string, string>? tags = null,
+        DateTime? timestamp = null
+        // , int stacklevel = 0 // Used for code locations
+    ) => Emit(MetricType.Distribution, key, value, unit, tags, timestamp);
+
+    /// <inheritdoc cref="IMetricAggregator.Set"/>
+    public void Set(
+        string key,
+        double value = 1.0,
+        MeasurementUnit? unit = null,
+        IDictionary<string, string>? tags = null,
+        DateTime? timestamp = null
+        // , int stacklevel = 0 // Used for code locations
+    ) => Emit(MetricType.Set, key, value, unit, tags, timestamp);
+
+    private void Emit(
+        MetricType type,
+        string key,
+        double value = 1.0,
+        MeasurementUnit? unit = null,
+        IDictionary<string, string>? tags = null,
+        DateTime? timestamp = null
+        // , int stacklevel = 0 // Used for code locations
+    )
+    {
+        timestamp ??= DateTime.UtcNow;
+        unit ??= MeasurementUnit.None;
+        var timeBucket = Buckets.GetOrAdd(
+            timestamp.Value.GetTimeBucketKey(),
+            _ => new ConcurrentDictionary<string, Metric>()
+        );
+
+        Func<string, Metric> addValuesFactory = type switch
+        {
+            MetricType.Counter => (string _) => new CounterMetric(key, value, unit.Value, tags, timestamp),
+            MetricType.Gauge => (string _) => new GaugeMetric(key, value, unit.Value, tags, timestamp),
+            MetricType.Distribution => (string _) => new DistributionMetric(key, value, unit.Value, tags, timestamp),
+            MetricType.Set => (string _) => new SetMetric(key, (int)value, unit.Value, tags, timestamp),
+            _ => throw new ArgumentOutOfRangeException(nameof(type), type, "Unknown MetricType")
+        };
+
+        timeBucket.AddOrUpdate(
+            GetMetricBucketKey(type, key, unit.Value, tags),
+            addValuesFactory,
+            (_, metric) =>
+            {
+                metric.Add(value);
```

Comment:
Might look like we have exclusive access to `metric` here but you're also reading from it to flush on another thread and while you're iterating on things there it might crash if mutate it here.

Reply:
Is that possible, given the [cutoff logic](https://github.com/getsentry/sentry-dotnet/blob/19dc74aa20d4d76fe6c94b3f78289b35517fe6af/src/Sentry/MetricAggregator.cs#L226-L225)?

This is a `ConcurrentDictionary` so all of the add/remove operations are atomic. The Flush logic will only retrieve buckets prior to the cuttoff (which means no mutations will have happened for `_flushShift` seconds. In an extreme case the flush shift could be set to a very small time value, in which case maybe at the end of a bucket a metric gets emitted and there's some chance of overlap?

I could create a lock, I guess, when flushing (so `metric.Add` operations would queue up while flushing). I think this is what the Python code does (although there's no ConcurrentDictionary in Python). 

Alternatively, the flushShift could be offset to guarantee all metric emits to have completed (e.g. 50-100ms). 

Thoughts? 🤔

Reply:
`ConcurrentDictionary` guarantees the callback on Add/Update won't run concurrently, but if you remove the reference (as it's done below) any access to the reference after is outside any locks. Now that it's `TryRemove` below (after the last commit) it might be OK, need to check the logic with iterating over the concurrentDic (which is safe), won't cause a situation where we find the key here (and run the update callback) while the iterator is alive with that reference. Does it block the `MoveNext()` call on the iterator?

I'll take another look (it's 10 PM on a Sunday) later on the IDE to see if I can spot anything but might be worth testing this with some code in a tight loop adding metrics and making the flush code run every 100ms or something to see if u can cause a crash. Let it run for a few hours for example.

Reply:
So turns out this was nothing to do with the Flush operation... the `ConcurrentDictionary<K,V>.AddOrUpdate` method is a bit misleading. See [here](https://github.com/getsentry/sentry-dotnet/pull/2949#issuecomment-1851463606).